azure/iac/storageaccounts.rego

```diff
@@ -32,22 +32,29 @@ azure_issue["storage_secure"] {
 
 storage_secure {
     lower(input.resources[_].type) == "microsoft.storage/storageaccounts"
-    not azure_attribute_absence["storage_secure"]
+    not azure_attribute_absence_old["storage_secure"]
+    not azure_issue["storage_secure"]
+}
+
+storage_secure {
+    azure_attribute_absence_new["storage_secure"]
```

Comment:
@shahinshirvani please remove `not azure_issue["storage_secure"]`. its not needed. if attribute is absent no need to check the value.

Reply:
@rezoan actually in this case, it is necessary
Because
```
azure_attribute_absence_new["storage_secure"] {
    resource := input.resources[_]
    lower(resource.type) == "microsoft.storage/storageaccounts"
    resource.apiVersion >= "2019-04-01"
    not resource.properties.supportsHttpsTrafficOnly
}
```
this ```not resource.properties.supportsHttpsTrafficOnly``` checks that ```supportsHttpsTrafficOnly``` isn't exist and the value of that is```false```
so ```azure_attribute_absence_new``` will true if ```supportsHttpsTrafficOnly``` exists and the value of that is false
and this is the opposite of policy
so we need the check the value

Reply:
@shahinshirvani 
this is not `or`. it is `and `you are checking inside `storage_secure `rule. so this rule will evaluate as "if `supportsHttpsTrafficOnly `**property does not exist** and at the same time its value is set to **true**". dead end right?

i have requested to remove `not azure_issue["storage_secure"]` from 
```
storage_secure {
    azure_attribute_absence_new["storage_secure"]
    not azure_issue["storage_secure"]
 }
```
which will make the rule to be passed if property does not exist (default true) and then create another one with 

```
storage_secure {
    not azure_attribute_absence_new["storage_secure"]
    not azure_issue["storage_secure"]
 }
```

Just make sure the property exist before checking its value.

Let me know if you still having issue.

Reply:
@rezoan
I got your point
but according the rules of OPA ```not resource.properties.supportsHttpsTrafficOnly```, check two things
first ```supportsHttpsTrafficOnly``` exists
second one  the value of ```supportsHttpsTrafficOnly``` is false
for more information please check this
![Screenshot (108)](https://user-images.githubusercontent.com/76029421/129450489-b374f184-1d80-4853-be64-80b2932fe45d.png)
https://www.openpolicyagent.org/docs/latest/policy-reference/#lookup

so assume this arm template:
```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "location": {
            "type": "string",
            "defaultValue": "[resourceGroup().location]",
            "metadata": {
                "description": "Specifies the Azure location where the key vault should be created."
            }
        },
        "storageAccountName": {
            "type": "string",
            "defaultValue": "[concat('storage', uniqueString(resourceGroup().id))]",
            "metadata": {
                "description": "Specifies the name of the storageAccount."
            }
        },
        "skuName": {
            "type": "string",
            "defaultValue": "Standard_LRS",
            "allowedValues": [
                "Premium_LRS",
                "Premium_ZRS",
                "Standard_GRS",
                "Standard_GZRS",
                "Standard_LRS",
                "Standard_RAGRS",
                "Standard_RAGZRS",
                "Standard_ZRS"
            ],
            "metadata": {
                "description": "Specifies the sku name for the storage account."
            }
        },
        "kind": {
            "type": "string",
            "defaultValue": "StorageV2",
            "allowedValues": [
                "BlobStorage",
                "BlockBlobStorage",
                "FileStorage",
                "Storage",
                "StorageV2"
            ],
            "metadata": {
                "description": "Specifies the type of storage account."
            }
        },
        "accessTier": {
            "type": "string",
            "defaultValue": "Hot",
            "allowedValues": [
                "Hot",
                "Cool"
            ],
            "metadata": {
                "description": "Required for storage accounts where kind = BlobStorage. The access tier used for billing."
            }
        },
        "minimumTlsVersion": {
            "type": "string",
            "defaultValue": "TLS1_2",
            "allowedValues": [
                "TLS1_0",
                "TLS1_1",
                "TLS1_2"
            ],
            "metadata": {
                "description": "Set the minimum TLS version to be permitted on requests to storage."
            }
        },
        "allowBlobPublicAccess": {
            "type": "bool",
            "defaultValue": true,
            "metadata": {
                "description": "Allow or disallow public access to all blobs or containers in the storage account."
            }
        },
        "allowSharedKeyAccess": {
            "type": "bool",
            "defaultValue": true,
            "metadata": {
                "description": "Indicates whether the storage account permits requests to be authorized with the account access key via Shared Key. If false, then all requests, including shared access signatures, must be authorized with Azure Active Directory (Azure AD)."
            }
        },
        "networkAclsBypass": {
            "type": "string",
            "defaultValue": "AzureServices",
            "allowedValues": [
                "AzureServices",
                "Logging",
                "Metrics",
                "None"
            ],
            "metadata": {
                "description": "Specifies whether traffic is bypassed by the indicated service."
            }
        },
        "networkAclsDefaultAction": {
            "type": "string",
            "defaultValue": "Deny",
            "allowedValues": [
                "Allow",
                "Deny"
            ],
            "metadata": {
                "description": "Specifies the default action of allow or deny when no other rules match."
            }
        }
    },
    "variables": {},
    "resources": [
        {
            "type": "Microsoft.Storage/storageAccounts",
            "apiVersion": "2021-01-01",
            "name": "[parameters('storageAccountName')]",
            "location": "[parameters('location')]",
            "sku": {
                "name": "[parameters('skuName')]"
            },
            "kind": "[parameters('kind')]",
            "properties": {
                "accessTier": "[parameters('accessTier')]",
                "minimumTlsVersion": "[parameters('minimumTlsVersion')]",
                "supportsHttpsTrafficOnly": false,
                "allowBlobPublicAccess": "[parameters('allowBlobPublicAccess')]",
                "allowSharedKeyAccess": "[parameters('allowSharedKeyAccess')]",
                "networkAcls": {
                    "bypass": "[parameters('networkAclsBypass')]",
                    "defaultAction": "[parameters('networkAclsDefaultAction')]"
                },
                "encryption": {
                    "services": {
                        "blob": {
                            "enabled": true,
                            "keyType": "string"
                        },
                        "file": {
                            "enabled": true,
                            "keyType": "string"
                        },
                        "table": {
                            "enabled": "boolean",
                            "keyType": "string"
                        },
                        "queue": {
                            "enabled": "boolean",
                            "keyType": "string"
                        }
                    },
                    "keySource": "Microsoft.Keyvault",
                    "requireInfrastructureEncryption": "boolean",
                    "keyvaultproperties": {
                        "keyname": "string",
                        "keyversion": "string",
                        "keyvaulturi": "string"
                    },
                    "identity": {
                        "userAssignedIdentity": "string"
                    }
                }
            }
        }
    ],
    "outputs": {
        "storageAccountName": {
            "type": "string",
            "value": "[parameters('storageAccountName')]"
        },
        "storageAccountResourceGroup": {
            "type": "string",
            "value": "[resourceGroup().name]"
        },
        "location": {
            "type": "string",
            "value": "[parameters('location')]"
        }
    }
}
```
in this arm template ```supportsHttpsTrafficOnly``` exists and the value of that is false
so this policy should be failed


but if we used this:
```
storage_secure {
    azure_attribute_absence_new["storage_secure"]
 }
```
```azure_attribute_absence_new``` will true , so ```storage_secure``` will be true and this is wrong

Reply:
@shahinshirvani good point but strange behavior. i somehow overlooked this behavior. 
not sure in OPA how this condition will validate the values other than Boolean type. some property has 'enabled'/'disabled' value which basically do the same and i believe the feature is not application for those.
Approving the PR.